grpc/stream.go

```diff
@@ -0,0 +1,341 @@
+package grpc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dop251/goja"
+	"github.com/grafana/xk6-grpc/lib/netext/grpcext"
+	"github.com/mstoykov/k6-taskqueue-lib/taskqueue"
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+type message struct {
+	msg []byte
+}
+
+type stream struct {
+	vu     modules.VU
+	client *Client
+
+	logger logrus.FieldLogger
+
+	methodDescriptor protoreflect.MethodDescriptor
+
+	method string
+	stream *grpcext.Stream
+
+	tagsAndMeta    *metrics.TagsAndMeta
+	tq             *taskqueue.TaskQueue
+	builtinMetrics *metrics.BuiltinMetrics
+	obj            *goja.Object // the object that is given to js to interact with the stream
+
+	closed bool
+	done   chan struct{}
+
+	writeQueueCh chan message
+
+	eventListeners *eventListeners
+}
+
+// defineStream defines the goja.Object that is given to js to interact with the Stream
+func defineStream(rt *goja.Runtime, s *stream) {
+	must(rt, s.obj.DefineDataProperty(
+		"on", rt.ToValue(s.on), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"write", rt.ToValue(s.write), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"end", rt.ToValue(s.end), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+}
+
+func (s *stream) beginStream() error {
+	req := &grpcext.StreamRequest{
+		Method:           s.method,
+		MethodDescriptor: s.methodDescriptor,
+	}
+
+	stream, err := s.client.conn.NewStream(s.vu.Context(), *req, metadata.MD{})
+	if err != nil {
+		return fmt.Errorf("failed to create a new stream: %w", err)
+	}
+	s.stream = stream
+
+	go s.loop()
+
+	return nil
+}
+
+func (s *stream) loop() {
+	readDataChan := make(chan map[string]interface{})
+
+	ctx := s.vu.Context()
+
+	defer func() {
+		ch := make(chan struct{})
+		s.tq.Queue(func() error {
+			defer close(ch)
+			return s.closeWithError(nil)
+		})
+
+		select {
+		case <-ch:
+		case <-ctx.Done():
+			// unfortunately it is really possible that k6 has been winding down the VU and the above code
+			// to close the connection will just never be called as the event loop no longer executes the callbacks.
+			// This ultimately needs a separate signal for when the eventloop will not execute anything after this.
+			// To try to prevent leaking goroutines here we will try to figure out if we need to close the `done`
+			// channel and wait a bit and close it if it isn't. This might be better off with more mutexes
+			timer := time.NewTimer(time.Millisecond * 250)
+			select {
+			case <-s.done:
+				// everything is fine
+			case <-timer.C:
+				close(s.done) // hopefully this means we won't double close
+			}
+			timer.Stop()
+		}
+
+		s.tq.Close()
+	}()
+
+	// read & write data from/to the stream
+	go s.readData(readDataChan)
+	go s.writeData()
+
+	ctxDone := ctx.Done()
+	for {
```

Comment:
Why not have this in the `readData` handler?

Reply:
Because the `readData` routine is responsible for reading the data from the stream, and this part of the code is responsible more for routing the data between handlers.

Reply:
>for routing the data between **handlers**

Do you mean listeners?

Then I think it could be more readable in a `listenerRouting` function. But if you prefer to have it here, it's fine for me.

Reply:
I prefer to keep it that way, at least for now